testing/ef-tests/src/cases/blockchain_test.rs

```diff
@@ -83,87 +88,178 @@ impl Case for BlockchainTestCase {
+            // Insert state hashes into the provider based on the expected state root.
+            let last_block = last_block.unwrap_or_default();
+            provider.insert_hashes(
+                0..=last_block.number,
+                last_block.hash(),
+                *expected_state_root,
+            )?;
+        }
+        _ => return Err(Error::MissingPostState),
+    }
+
+    // Drop the provider without committing to the database.
+    drop(provider);
+    Ok(())
+}
+
+fn execute_blocks<
+    Provider: DBProvider
+        + StateWriter<Receipt = reth_primitives::Receipt>
+        + BlockHashReader
+        + StateCommitmentProvider,
+>(
+    provider: &Provider,
+    blocks_without_genesis: &[RecoveredBlock<Block<TransactionSigned>>],
+    chain_spec: Arc<ChainSpec>,
+) -> Result<Vec<ExecutionWitness>, Error> {
+    let executor_provider = EthExecutorProvider::ethereum(chain_spec);
+
+    let mut exec_witnesses = Vec::new();
+
+    for block in blocks_without_genesis {
+        let state_provider = LatestStateProviderRef::new(provider);
+        let state_db = StateProviderDatabase(&state_provider);
+        let block_executor = executor_provider.executor(state_db);
+
+        let mut witness_record = ExecutionWitnessRecord::default();
+
+        let output = block_executor
+            .execute_with_state_closure(&(*block).clone(), |statedb: &State<_>| {
+                witness_record.record_executed_state(statedb);
+            })
+            .map_err(|_| Error::BlockExecutionFailed)?;
+
+        provider.write_state(
+            &ExecutionOutcome::single(block.number, output),
+            OriginalValuesKnown::Yes,
+            StorageLocation::Database,
+        )?;
+
+        let ExecutionWitnessRecord { hashed_state, codes, keys } = witness_record;
+        let state = state_provider.witness(Default::default(), hashed_state)?;
+        let exec_witness = ExecutionWitness { state, codes, keys };
+        exec_witnesses.push(exec_witness);
+    }
+
+    Ok(exec_witnesses)
+}
+
+fn compute_stateless_input(
+    execution_witnesses: Vec<ExecutionWitness>,
+    blocks_with_genesis: &[RecoveredBlock<Block<TransactionSigned>>],
+    chain_spec: Arc<ChainSpec>,
+) -> Vec<Input> {
+    let mut guest_inputs = Vec::new();
+
+    assert_eq!(blocks_with_genesis.len() - 1, execution_witnesses.len());
+
+    for (current_block, execution_witness) in
+        blocks_with_genesis.iter().skip(1).zip(execution_witnesses)
+    {
+        // Skip the genesis block, it has no ancestors and no execution
+        // TODO: This should ideally look inside of the block and check for blockhash opcodes
+        // TODO: to see how many ancestors we need instead of just getting all of the previous
+        // TODO: blocks
+        let ancestors = blocks_with_genesis[0..current_block.number as usize]
+            .iter()
+            .map(|block| block.header())
+            .cloned()
+            .collect();
```

Comment:
This should be getting the last 256 blocks walking back from the current, not from 0